src/legacy/huffman.rs

```diff
@@ -0,0 +1,260 @@
+
+            sym_idx[l] = sym_idx[l - 1] + count[l - 1];
+            self.offset_first_sym_idx[l] = sym_idx[l].wrapping_sub(code[l]);
+        }
+
+        // Build mapping from index to symbol and populate the lookup table.
+        lengths
+            .iter()
+            .enumerate()
+            .take(n)
+            .for_each(|(i, code_len)| {
+                let l = *code_len as usize;
+                if l == 0 {
+                    return;
+                }
+
+                self.syms[sym_idx[l] as usize] = i as u16;
+                sym_idx[l] += 1;
+
+                if l <= HUFFMAN_LOOKUP_TABLE_BITS as usize {
+                    self.table_insert(i, l, code[l]);
+                    code[l] += 1;
+                }
+            });
+
+        Ok(())
+    }
+
+    pub fn table_insert(&mut self, sym: usize, len: usize, codeword: u16) {
+        debug_assert!(len <= HUFFMAN_LOOKUP_TABLE_BITS as usize);
+
+        let codeword = reverse_lsb(codeword, len); // Make it LSB-first.
+        let pad_len = HUFFMAN_LOOKUP_TABLE_BITS as usize - len;
+
+        // Pad the pad_len upper bits with all bit combinations.
+        for padding in 0..(1 << pad_len) {
+            let index = (codeword | (padding << len)) as usize;
+            debug_assert!(sym <= u16::MAX as usize);
+            self.table[index].sym = sym as u16;
+            debug_assert!(len <= u8::MAX as usize);
+            self.table[index].len = len as u8;
+        }
+    }
+
+    /// Use the decoder d to decode a symbol from the LSB-first zero-padded bits.
+    /// Returns the decoded symbol number or an error if no symbol could be decoded.
+    /// *num_used_bits will be set to the number of bits used to decode the symbol,
+    /// or zero if no symbol could be decoded.
+    pub fn huffman_decode<T: std::io::Read + Seek, E: Endianness>(
+        &mut self,
+        length: u64,
+        is: &mut BitReader<T, E>,
+    ) -> std::io::Result<u16> {
+        // First try the lookup table.
+        let read_bits1 = (HUFFMAN_LOOKUP_TABLE_BITS as u64).min(length - is.position_in_bits()?);
+        let lookup_bits = !is.read::<u8>(read_bits1 as u32)? as usize;
+        debug_assert!(lookup_bits < self.table.len());
+        if self.table[lookup_bits].len != 0 {
+            debug_assert!(self.table[lookup_bits].len <= HUFFMAN_LOOKUP_TABLE_BITS);
+            is.seek_bits(io::SeekFrom::Current(
+                -(read_bits1 as i64) + self.table[lookup_bits].len as i64,
+            ))?;
+            return Ok(self.table[lookup_bits].sym);
+        }
+
+        // Then do canonical decoding with the bits in MSB-first order.
+        let read_bits2 = (HUFFMAN_LOOKUP_TABLE_BITS as u64).min(length - is.position_in_bits()?);
+        let mut bits = reverse_lsb(
+            (lookup_bits | ((!is.read::<u8>(read_bits2 as u32)? as usize) << read_bits1)) as u16,
+            MAX_HUFFMAN_BITS,
+        );
+
+        for l in HUFFMAN_LOOKUP_TABLE_BITS as usize + 1..=MAX_HUFFMAN_BITS {
```

Comment:
Add clarifying parentheses around `HUFFMAN_LOOKUP_TABLE_BITS as usize + 1`; otherwise people new to Rust would assume `1 + 2..5` meant `3..6`.